components/file_system/src/rate_limiter.rs

```diff
@@ -1,148 +1,558 @@
             }
         }
     }
 
-    pub fn fetch(&self, io_type: IOType, io_op: IOOp) -> usize {
-        match io_op {
-            IOOp::Read => self.read[io_type as usize].load(Ordering::Relaxed),
-            IOOp::Write => self.write[io_type as usize].load(Ordering::Relaxed),
+    pub fn reset(&self) {
+        for i in 0..IOType::COUNT {
+            self.read_bytes[i].store(0, Ordering::Relaxed);
+            self.write_bytes[i].store(0, Ordering::Relaxed);
+        }
+    }
+}
+
+macro_rules! do_sleep {
+    ($duration:expr, "sync") => {
+        std::thread::sleep($duration)
+    };
+    ($duration:expr, "async") => {
+        tokio::time::delay_for($duration).await
+    };
+}
+
+const DEFAULT_REFILL_PERIOD: Duration = Duration::from_millis(40);
+
+/// Limit total IO flow below provided threshold by throttling lower-priority IOs.
+/// Rate limit is disabled when total IO threshold is set to zero.
+#[derive(Debug)]
+struct RawIORateLimiter {
+    // IO amount passed through within current epoch
+    ios_through: [CachePadded<AtomicUsize>; IOPriority::COUNT],
+    // Maximum IOs permitted within current epoch
+    ios_per_epoch: [CachePadded<AtomicUsize>; IOPriority::COUNT],
+    protected: Mutex<RawIORateLimiterProtected>,
+}
+
+#[derive(Debug)]
+struct RawIORateLimiterProtected {
+    next_refill_time: Instant,
+    // IO amount that is drew from the next epoch in advance
+    pending_ios: [usize; IOPriority::COUNT],
+    // estimated throughput of recent epochs
+    estimated_ios_through: [IOThroughputEstimator; IOPriority::COUNT],
+}
+
+impl RawIORateLimiterProtected {
+    fn new() -> Self {
+        RawIORateLimiterProtected {
+            next_refill_time: Instant::now_coarse() + DEFAULT_REFILL_PERIOD,
+            pending_ios: [0; IOPriority::COUNT],
+            estimated_ios_through: [IOThroughputEstimator::new(); IOPriority::COUNT],
+        }
+    }
+}
+
+#[derive(Debug, Copy, Clone)]
+struct IOThroughputEstimator {
+    /// total count of sampled epochs
+    count: usize,
+    /// sum of IOs amount over a window
+    sum: usize,
+}
+
+impl IOThroughputEstimator {
+    fn new() -> IOThroughputEstimator {
+        IOThroughputEstimator { count: 0, sum: 0 }
+    }
+
+    fn maybe_update_estimation(&mut self, v: usize) -> Option<usize> {
+        const WINDOW_SIZE: usize = 5;
+        self.count += 1;
+        self.sum += v;
+        // average over DEFAULT_REFILL_PERIOD*WINDOW_SIZE=200ms window
+        if self.count % WINDOW_SIZE == 0 {
+            let avg = self.sum / WINDOW_SIZE;
+            self.sum = 0;
+            Some(avg)
+        } else {
+            None
+        }
+    }
+}
+
+/// Actual implementation for requesting IOs from RawIORateLimiter.
+/// An attempt will be recorded first. If the attempted amount exceeds the available quotas of
+/// current epoch, the requester will register itself and sleep until next epoch.
+macro_rules! request_imp {
```

Comment:
why can't make it a funcation?

Reply:
Because a function can't be both synchronous and asynchronous, macro is the only way to dedup these code.